crates/stages/src/stages/headers.rs

```diff
@@ -54,56 +67,119 @@ where
     Downloader: HeaderDownloader,
 {
     /// Create a new header stage
-    pub fn new(database: Provider, downloader: Downloader, mode: HeaderSyncMode) -> Self {
-        Self { provider: database, downloader, mode, sync_gap: None, buffer: None }
-    }
-
-    fn is_stage_done<DB: Database>(
-        &self,
-        tx: &<DB as Database>::TXMut,
-        checkpoint: u64,
-    ) -> Result<bool, StageError> {
-        let mut header_cursor = tx.cursor_read::<tables::CanonicalHeaders>()?;
-        let (head_num, _) = header_cursor
-            .seek_exact(checkpoint)?
-            .ok_or_else(|| ProviderError::HeaderNotFound(checkpoint.into()))?;
-        // Check if the next entry is congruent
-        Ok(header_cursor.next()?.map(|(next_num, _)| head_num + 1 == next_num).unwrap_or_default())
+    pub fn new(
+        database: Provider,
+        downloader: Downloader,
+        mode: HeaderSyncMode,
+        consensus: Arc<dyn Consensus>,
+    ) -> Self {
+        Self {
+            provider: database,
+            downloader,
+            mode,
+            consensus,
+            sync_gap: None,
+            hash_collector: Collector::new(100 * (1024 * 1024)).unwrap(),
+            header_collector: Collector::new(100 * (1024 * 1024)).unwrap(),
+            is_etl_ready: false,
+        }
     }
 
-    /// Write downloaded headers to the given transaction
+    /// Write downloaded headers to the given transaction from ETL.
     ///
-    /// Note: this writes the headers with rising block numbers.
+    /// Writes to the following tables:
+    /// [`tables::Headers`],[`tables::CanonicalHeaders`],[`tables::HeaderTD`] and
+    /// [`tables::HeaderNumbers`].
     fn write_headers<DB: Database>(
-        &self,
+        &mut self,
         tx: &<DB as Database>::TXMut,
-        headers: Vec<SealedHeader>,
-    ) -> Result<Option<BlockNumber>, StageError> {
-        trace!(target: "sync::stages::headers", len = headers.len(), "writing headers");
-
-        let mut cursor_header = tx.cursor_write::<tables::Headers>()?;
-        let mut cursor_canonical = tx.cursor_write::<tables::CanonicalHeaders>()?;
-
-        let mut latest = None;
-        // Since the headers were returned in descending order,
-        // iterate them in the reverse order
-        for header in headers.into_iter().rev() {
+    ) -> Result<BlockNumber, StageError> {
+        trace!(target: "sync::stages::headers", len = self.header_collector.len(), "writing headers");
```

Comment:
i think we need some kind of info logs here, because we only write them during the download phase, and being silent during the ETL phase